data/completion/complete.sh

```diff
@@ -0,0 +1,73 @@
+# -*- bash -*-
+
+# _complete_from_snap serialises the tab completion request and sends it off to
+# the appropriate 'snap run --command=complete', and de-serialises the response
+# into the usual tab completion result.
+_complete_from_snap() {
+    {
+        read -a opts
+        # opts is expected to be a series of compopt options
+        if [[ ${#opts[@]} -gt 0 ]]; then
+            if [[ "${opts[0]}" == "cannot" ]]; then
+                # older snap-execs sent errors over stdout :-(
+                return 1
+            fi
+
+            for i in "${opts[@]}"; do
+                if ! [[ "$i" =~ ^[a-z]+$ ]]; then
+                    # non-alphanumeric option; something awry
+                    return 2
+                fi
+            done
+        fi
+
+        read bounced
+        case "$bounced" in
+            ""|"alias"|"export"|"job"|"variable")
+                # OK
+                ;;
+            *)
+                # unrecognised bounce
+                return 2
+                ;;
+        esac
+
+        read sep
+        if [ "$sep" ]; then
+            # non-blank separator? madness!
+            return 2
+        fi
+        local oldIFS="$IFS"
+
+        if [ ! "$bounced" ]; then
+            local IFS=$'\n'
+            COMPREPLY=( $( \grep -v '[[:cntrl:];?*{}]' ) )
+            IFS="$oldIFS"
+        fi
+
+        if [[ ${#opts[@]} -gt 0 ]]; then
+            compopt $(printf " -o %s" "${opts[@]}")
+        fi
+        if [ "$bounced" ]; then
+            COMPREPLY+=(compgen -A "$bounced" -- "${COMP_WORDS[$COMP_CWORD]}")
```

Comment:
Regarding quoting unilaterally, I agree: it really depends on what the completer script is doing on whether or not it should be heavily quoted. Your observation about _filedir is keen-- nice! I think that with our grep, we are ok (but see inline comment on the bounce case) and fixing _filedir is a bug fix.